### PR TITLE
test: add cluster mesh conformance tests with Kind 

### LINKED ISSUE #9994: Validate multicluster + ipsec + tunnel
We should figure out some automation around testing this combination and validate the functionality against a test suite for encryption in conjunction with tunneling mode, and ensure this works across clusters.

### LINKED ISSUE #23322: Enhancement: Kind-based testing and better test coverage for Cluster Mesh
This issue tracks the series of test enhancements of Cluster Mesh. This can be divided into several tasks.

**Migrate the current GCP-based testing infrastructure to a Kind-based infrastructure for better speed and flakiness**
At this point, we don't need to add any test cases. Only Kind migration should be done.

**Make a feature matrix and test scenarios for them (like a [datapath test](https://github.com/cilium/cilium/blob/master/.github/workflows/conformance-datapath.yaml#L123-L185))**
From Cluster Mesh's perspective, inter-cluster communication is the same as inter-node communication, so the datapath test covers most basic connectivity functionalities. So, we should focus on testing the Global / Shared service under some conditions. My suggestion would be like this.
  - encryption: "none" | "ipsec" | "wireguard"
  - tunnel: "disabled" | "vxlan"
  - ipFamily: "ipv4-only" | "dual-stack" | "ipv6-only"
  
Here are some issues that led me to create this issue.
- We couldn’t notice that Cluster Mesh was not working at all in IPv6 single stack environment (https://github.com/cilium/cilium/pull/22962)
- Wireguard test is somehow very flaky in GCP (It's not the matter of WireGuard test I think. I could complete the test in my local with Kind. Ref: https://github.com/cilium/cilium/pull/23045)

I have [partially done the Kind migration](https://github.com/cilium/cilium/compare/master...YutaroHayakawa:cilium:yutaro/oss/clustermesh-tests
). Please use this as an example. I currently use nektos/act for local testing, but it may not be accepted on upstream given the context in https://github.com/cilium/cilium/pull/20118. If it is still the case, let's remove act part on the actual PR (it is still useful for local testing, I think).

This issue can be closed once all the above tasks are achieved.

### PR DESCRIPTION
**Pull request by giorio94**
This PR introduces the new kind-based conformance tests for cluster mesh, implementing a test matrix to validate (a subset of) the following combinations:
* encryption: "none" | "ipsec" | "wireguard"
* tunnel: "disabled" | "vxlan"
* ipfamily: "ipv4-only" | "dual-stack" | ipv6-only"
* kube-proxy: "iptables" | "kpr"

Additional context can be found in the enhancement proposal: https://github.com/cilium/cilium/issues/23322

Fixes: #23322
Fixes: #9994

Co-authored-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>
Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

```release-note
test: add cluster mesh conformance tests with Kind 
```

### CODE REVIEW

File: .github/workflows/conformance-clustermesh.yaml
Comment (sayboras) [inline .github/workflows/conformance-clustermesh.yaml]: wait for second cluster as well ? or we can just remove the hubble for second cluster, as the connectivity tests are mainly in cluster 1 only.

File: .github/workflows/conformance-clustermesh.yaml
Comment (sayboras) [inline .github/workflows/conformance-clustermesh.yaml:429]: nit: just to align with previous commands

```suggestion
        cilium clustermesh status --wait --context ${{ env.contextName1 }}
        cilium clustermesh status --wait --context ${{ env.contextName2 }}
 ```

File: .github/workflows/conformance-clustermesh.yaml
Comment (sayboras) [inline .github/workflows/conformance-clustermesh.yaml:440]: same as above

File: .github/workflows/conformance-clustermesh.yaml
Comment (sayboras) [inline .github/workflows/conformance-clustermesh.yaml]: should we pass the context here to avoid any potential confusion?

**Review by sayboras — APPROVED** [review]
This is awesome :heavy_check_mark:

File: .github/workflows/conformance-clustermesh.yaml
Comment (giorio94) [reply .github/workflows/conformance-clustermesh.yaml:429]: I've moved the context to be the first parameter of all commands, for consistency

File: .github/workflows/conformance-clustermesh.yaml
Comment (giorio94) [reply .github/workflows/conformance-clustermesh.yaml]: I've added the wait command also for the second cluster. I would personally keep hubble enabled in both clusters, given the multicluster tests, although the relay is present only in the first one.

File: .github/workflows/conformance-clustermesh.yaml
Comment (sayboras) [inline .github/workflows/conformance-clustermesh.yaml]: any reason to plan running this job with pull_request_target ? AFAIK, there is no secret used in this workflow :thinking:

Comment (giorio94) [reply .github/workflows/conformance-clustermesh.yaml]: Actually the idea was keep the same approach as the existing one, and run the tests on `issue_command` (and schedule) only. This is consistent with the approach adopted also in the conformance-datapath workflow, as both spawn a quite large number of jobs. I'm obviously open to change the approach if you feel running on pull_request would be better.

**Review by pchaigno — APPROVED** [review]
LGTM assuming the second commit is removed.

Is the plan to remove the existing clustermesh workflow once this one is deemed stable?

File: .github/workflows/conformance-clustermesh.yaml
Comment (sayboras) [reply .github/workflows/conformance-clustermesh.yaml]: thanks for clarification :100:

**Comment by giorio94** [comment]
> Is the plan to remove the existing clustermesh workflow once this one is deemed stable?

Personally I would do so. I wonder if it is better to remove that in this PR or in a later one.

**Comment by pchaigno** [comment]
I would remove in a subsequent PR after giving it a bit of time. New tests are often more flaky than existing one so maybe best to wait a bit and see.

File: .github/workflows/conformance-clustermesh.yaml
Comment (YutaroHayakawa) [inline .github/workflows/conformance-clustermesh.yaml:59]: Maybe better to change the trigger phrase.

File: .github/workflows/conformance-clustermesh.yaml
Comment (YutaroHayakawa) [inline .github/workflows/conformance-clustermesh.yaml:80]: Same as above.

File: .github/workflows/conformance-clustermesh.yaml
Comment (YutaroHayakawa) [inline .github/workflows/conformance-clustermesh.yaml:120]: Same as above.

File: .github/workflows/conformance-clustermesh.yaml
Comment (tklauser) [inline .github/workflows/conformance-clustermesh.yaml:22]: I think this comment no longer applies and can be removed now that the workflow is triggered `on.pull_request`.

File: .github/workflows/conformance-clustermesh.yaml
Comment (YutaroHayakawa) [inline .github/workflows/conformance-clustermesh.yaml]: It would be nice to have trigger phrases written in here like other workflows.

https://github.com/cilium/cilium/blob/master/.github/workflows/conformance-multicluster.yaml#L1

Comment (tklauser) [inline .github/workflows/conformance-clustermesh.yaml]: Let's add the trigger phrase to the test name so it's easier to find from a PR's actions status overview.

```suggestion
name: ClusterMesh (ci-clustermesh)
```

File: .github/workflows/conformance-clustermesh.yaml
Comment (tklauser) [inline .github/workflows/conformance-clustermesh.yaml:59]: This is using the same trigger phrase as the existing `conformance-multicluster.yaml` test. I assume this is fine given we eventually want to remove the latter (ref. https://github.com/cilium/cilium/pull/23496#issuecomment-1439931820)?

**Review by tklauser — APPROVED** [review]
LGTM, some small nits inline.

File: .github/workflows/conformance-clustermesh.yaml
Comment (giorio94) [reply .github/workflows/conformance-clustermesh.yaml:22]: The `on.pull_request` trigger is planned to be removed (it is currently uncommented in the temporary commit). I've kept it for the moment just for ease of testing.

File: .github/workflows/conformance-clustermesh.yaml
Comment (YutaroHayakawa) [inline .github/workflows/conformance-clustermesh.yaml:275]: `--apiserver-version` is not a parameter to specify a tag. You should do `--apiserver-image=quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/clustermesh-apiserver-ci:${SHA}`.

File: .github/workflows/conformance-clustermesh.yaml
Comment (YutaroHayakawa) [inline .github/workflows/conformance-clustermesh.yaml]: `--relay-version` is not a parameter to specify a tag. You should do `--relay-image=quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/hubble-relay-ci:${SHA}`.

**Review by YutaroHayakawa — CHANGES_REQUESTED** [review]
Great job! Overall, looks great, but some small points. Especially the one of clustermesh-apiserver installation process leads to pick a wrong image in the test IIUC, so should be fixed.

**Comment by YutaroHayakawa** [comment]
Oh, Tobias was reviewing at the same time, and some comments are duplicated. Sorry 😓

File: .github/workflows/conformance-clustermesh.yaml
Comment (tklauser) [reply .github/workflows/conformance-clustermesh.yaml:22]: :facepalm: oh, I didn't spot the test commit. All good in that case.

File: .github/workflows/conformance-clustermesh.yaml
Comment (pchaigno) [reply .github/workflows/conformance-clustermesh.yaml]: We should add some linter for this. I always forget.

File: .github/workflows/conformance-clustermesh.yaml
Comment (pchaigno) [reply .github/workflows/conformance-clustermesh.yaml:59]: Why? If the old test runs, I'd expect us to want to run the new one as well.

File: .github/workflows/conformance-clustermesh.yaml
Comment (aanm) [inline .github/workflows/conformance-clustermesh.yaml]: same here

File: .github/workflows/conformance-clustermesh.yaml
Comment (aanm) [inline .github/workflows/conformance-clustermesh.yaml]: please add a comment with the version of this commit so that renovate can update this dependency automatically. Similar to what is done on line 330

**Review by aanm — CHANGES_REQUESTED** [review]

File: .github/workflows/conformance-clustermesh.yaml
Comment (giorio94) [reply .github/workflows/conformance-clustermesh.yaml]: Missed that, sorry

File: .github/workflows/conformance-clustermesh.yaml
Comment (giorio94) [reply .github/workflows/conformance-clustermesh.yaml:275]: For the clustermesh API server image it is actually, since it is not yet using helm under the hood: https://github.com/cilium/cilium-cli/blob/fc94d2015cf9d33f087c53760dfd8f3aefffd8de/internal/utils/utils.go#L78-L80

This is also confirmed by the output of the installation command: https://github.com/cilium/cilium/actions/runs/4235266535/jobs/7358697796#step:16:18

File: .github/workflows/conformance-clustermesh.yaml
Comment (giorio94) [reply .github/workflows/conformance-clustermesh.yaml]: That flag used indeed, thanks! I've changed that to use `--helm-set` explicitly, avoiding the deprecated flags.

**Comment by giorio94** [comment]
I've force-pushed to include the review suggestions. I've kept `/ci-multicluster` as trigger phrase since this is expected to eventually replace the existing multicluster tests.

Link to passing tests: https://github.com/cilium/cilium/actions/runs/4250727144/jobs/7392185537

**Review by aanm — APPROVED** [review]

**Comment by giorio94** [comment]
Last push bumped the Cilium CLI version to the newly released 1.13.0, and removed the associated temporary fixes.

Full diff:
```diff
--- a/.github/workflows/conformance-clustermesh.yaml
+++ b/.github/workflows/conformance-clustermesh.yaml
@@ -66,7 +66,7 @@ concurrency:
 env:
   kind_version: v0.17.0
   # renovate: datasource=github-releases depName=cilium/cilium-cli
-  cilium_cli_version: v0.12.13
+  cilium_cli_version: v0.13.0
   clusterName1: cluster1-${{ github.run_id }}
   clusterName2: cluster2-${{ github.run_id }}
   contextName1: kind-cluster1-${{ github.run_id }}
@@ -287,22 +287,6 @@ jobs:
             --test='!/pod-to-cidr'"
         fi
 
-        # Skip client-ingress/client-to-client tests in IPv6-only clusters, as the network policy
-        # currently allows only IPv4 ICMP Echo Requests, while blocks IPv6 ICMP Echo Requests
-        # TODO: Remove once cilium/cilium-cli#1403 gets merged and released.
-        if [[ "${{ matrix.ipFamily }}" == "ipv6" ]]; then
-          CONNECTIVITY_TEST_DEFAULTS="$CONNECTIVITY_TEST_DEFAULTS \
-            --test='!client-ingress-icmp/client-to-client' \
-            --test='!client-ingress-from-other-client-icmp-deny/client-to-client'"
-        fi
-
-        # Skip node-port tests if kube-proxy-replacement is disabled, as not working (#23266)
-        # TODO: Remove once cilium/cilium-cli#1401 gets merged and released.
-        if [[ "${{ matrix.kube-proxy }}" == "iptables" ]]; then
-          CONNECTIVITY_TEST_DEFAULTS="$CONNECTIVITY_TEST_DEFAULTS \
-            --test='!/pod-to-.*-nodeport'"
-        fi
-
         echo cilium_install_defaults="${CILIUM_INSTALL_DEFAULTS} ${CILIUM_INSTALL_TUNNEL} ${CILIUM_INSTALL_IPFAMILY}" >> $GITHUB_OUTPUT
         echo hubble_enable_defaults=${HUBBLE_ENABLE_DEFAULTS} >> $GITHUB_OUTPUT
         echo connectivity_test_defaults=${CONNECTIVITY_TEST_DEFAULTS} >> $GITHUB_OUTPUT
```

**Comment by giorio94** [comment]
There's apparently an issue with IPv6 tests, I'm investigating what has happened.

**Comment by giorio94** [comment]
I've rolled back the CLI bump due to the above mentioned issue. Fix: https://github.com/cilium/cilium-cli/pull/1414

**Comment by giorio94** [comment]
Manually rerunning one test which failed due to https://github.com/cilium/cilium/issues/24021

**Review by YutaroHayakawa — APPROVED** [review]
Sorry for blocking this. Now, LGTM 👍

**Comment by giorio94** [comment]
Given that all requested reviews are in, I've dropped the temporary commit used to run the tests.
Link to last successful workflow run: https://github.com/cilium/cilium/actions/runs/4264336582/jobs/7423694477

One final doubt before marking ready to merge: once merged, renovate will attempt to bump the Cilium CLI version to 1.13, but IPv6 tests would then fail. How should we handle that?

**Comment by giorio94** [comment]
> IIUC, the IPv6 tests were fixed by our PR https://github.com/cilium/cilium-cli/pull/1414. If yes, we could tag cilium-cli v0.13.1 with that fix included.

@tklauser IMHO Tagging cilium-cli v0.13.1 with the fix would be the best solution, if possible. Then, I could also drop the workarounds which are currently present due to failing tests in IPv6-only clusters. Otherwise I'll drop the renovate comment.

**Comment by giorio94** [comment]
Last push bumped the Cilium CLI version to the newly released 1.13.1, and removed the associated temporary fixes.

**Comment by giorio94** [comment]
Changes since last reviews:

* I've disabled flow validation through hubble, for consistency with the other existing tests, since that almost doubled tests duration. The context flag is removed since redundant (it is also specified in the actual invocation).

```diff
--- a/.github/workflows/conformance-clustermesh.yaml
+++ b/.github/workflows/conformance-clustermesh.yaml
@@ -274,7 +274,8 @@ jobs:
         CLUSTERMESH_ENABLE_DEFAULTS="--apiserver-image=quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/clustermesh-apiserver-ci \
           --apiserver-version=${SHA} --service-type=NodePort"
 
-        CONNECTIVITY_TEST_DEFAULTS="--context=${{ env.contextName1 }} \
+        CONNECTIVITY_TEST_DEFAULTS="--hubble=false \
+          --flow-validation=disabled \
           --multi-cluster=${{ env.contextName2 }} \
           --external-target=google.com \
           --collect-sysdump-on-failure"
```

* I've explicitly configured the coredns upstream DNS servers since, in case of IPv4+IPV6 clusters, the default ones included also IPv6 upstreams, which are not reachable from the GH actions environments, and caused spurious failures. 

```diff
--- a/.github/workflows/conformance-clustermesh.yaml
+++ b/.github/workflows/conformance-clustermesh.yaml
         config: ./.github/kind-config-cluster2.yaml
         wait: 0 # The control-plane never becomes ready, since no CNI is present
 
+    # Make sure that coredns uses IPv4-only upstream DNS servers also in case of clusters
+    # with IP family dual, since IPv6 ones are not reachable and cause spurious failures.
+    - name: Configure the coredns nameservers
+      if: matrix.ipfamily == 'dual'
+      run: |
+        COREDNS_PATCH="
+        spec:
+          template:
+            spec:
+              dnsPolicy: None
+              dnsConfig:
+                nameservers:
+                - 8.8.4.4
+                - 8.8.8.8
+        "
+
+        kubectl --context ${{ env.contextName1 }} patch deployment -n kube-system coredns --patch="$COREDNS_PATCH"
+        kubectl --context ${{ env.contextName2 }} patch deployment -n kube-system coredns --patch="$COREDNS_PATCH"
+
     - name: Wait for images to be available
       timeout-minutes: 10
       shell: bash
```

Link to successful run: https://github.com/cilium/cilium/actions/runs/4342212579

@tklauser, @pchaigno, @sayboras, @aanm, @YutaroHayakawa Please, have a final look if you'd like.  From my side, then, this is ready to be marked as ready to merge.

**Comment by giorio94** [comment]
Ingress related tests failed due to known flake https://github.com/cilium/cilium/issues/23960. I'm not re-running them, since this PR does not introduce any code changes.

**Comment by sayboras** [comment]
> Ingress related tests failed due to known flake #23960. I'm not re-running them, since this PR does not introduce any code changes.

It should be fixed in master now, but we don't need to re-run it here.

**Review by sayboras — APPROVED** [review]
Still LGTM :heavy_check_mark:, this is good stuff :astonished:

**Comment by giorio94** [comment]
All required reviews are in, and test failures are unrelated, since this PR is only introducing new tests. 
Marking as ready to merge.

**Comment by sayboras** [comment]
Looks good to me, the issue with Ingress GHA failure is already addressed in master branch.

**Comment by michi-covalent** [comment]
do we want to backport this to stable branches? 👀

**Pull request merged**
